### PR TITLE
FEAT-#2664: Add `to_labels` algebraic operator

### LINKED ISSUE #2664: Algebraic operator: `to_labels`
Important for setting labels from a column like `set_index`.

### PR DESCRIPTION
**Pull request by devin-petersohn**
Resolves #2664

This add the algebraic operator for `to_labels`, which enables Modin to
better optimize the movement of data to metadata. See more in the paper
about the algebraic operator:
http://www.vldb.org/pvldb/vol13/p2033-petersohn.pdf

Signed-off-by: Devin Petersohn <devin.petersohn@gmail.com>

## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #2664 
- [x] tests ~added and~ passing

### CODE REVIEW

**Comment by devin-petersohn** [comment]
This does not include any performance enhancements at the moment, because we synchronously wait for the labels to be collected before moving on. A subsequent PR will have to be made if we want the materialization of the row labels to happen asynchronously.

File: modin/backends/pandas/query_compiler.py [hunk 1]
```diff
@@ -546,6 +548,56 @@ def reset_index(self, **kwargs):
...
+            # Appending keeps the original order of the index levels, then appends the
+            # new index objects.
+            names = list(self.index.names)
+            if isinstance(self.index, pandas.MultiIndex):
+                for i in range(self.index.nlevels):
+                    arrays.append(self.index._get_level_values(i))
+            else:
+                arrays.append(self.index)
+
+            # Add the names in the correct order
```
Comment (williamma12) [inline modin/backends/pandas/query_compiler.py:583]:
```suggestion
            # Add the names in the correct order.
```

File: modin/engines/base/frame/data.py [hunk 2]
```diff
@@ -536,6 +537,30 @@ def mask(
             row_numeric_idx=new_row_order, col_numeric_idx=new_col_order
         )
 
+    def to_labels(self, column_list: List[Hashable]) -> "BasePandasFrame":
+        """Move one or more columns into the row labels. Previous labels are dropped.
+
+        Parameters
+        ----------
+        column_list : list of hashable
+            The list of column names to place as the new row labels
```
Comment (williamma12) [inline modin/engines/base/frame/data.py:546]:
```suggestion
            The list of column names to place as the new row labels.
```

File: modin/engines/base/frame/data.py [hunk 3]
```diff
@@ -536,6 +537,30 @@ def mask(
...
+
+        Returns
+        -------
+        BasePandasFrame
+            A new BasePandasFrame that has the updated labels
```
Comment (williamma12) [inline modin/engines/base/frame/data.py:551]:
```suggestion
            A new BasePandasFrame that has the updated labels.
```

File: modin/backends/base/query_compiler.py [hunk 4]
```diff
@@ -484,6 +485,30 @@ def reset_index(self, **kwargs):
...
+        self, keys: List[Hashable], drop: bool = True, append: bool = False
+    ):
+        """Create new row labels from a list of columns.
+
+        Parameters
+        ----------
+        keys : list of hashable
+            The list of column names that will become the new index.
+        drop : boolean
+            Whether or not to drop the columns provided in the `keys` argument
```
Comment (williamma12) [inline modin/backends/base/query_compiler.py:498]:
```suggestion
            Whether or not to drop the columns provided in the `keys` argument.
```

**Review by williamma12 — CHANGES_REQUESTED** [review]
Looks good! Just very minor missing periods

**Review by williamma12 — APPROVED** [review]

**Review by williamma12 — APPROVED** [review]

**Pull request merged**